compiler/src/dmd/funcsem.d

```diff
@@ -573,666 +559,691 @@ void funcDeclarationSemantic(Scope* sc, FuncDeclaration funcdecl)
 
-        // if static function, do not put in vtbl[]
-        if (!funcdecl.isVirtual())
+    /* Purity and safety can be inferred for some functions by examining
+     * the function body.
+     */
+    if (funcdecl.canInferAttributes(sc))
+        funcdecl.initInferAttributes();
+
+    funcdecl.semanticRun = PASS.semanticdone;
+
+    /* Save scope for possible later use (if we need the
+     * function internals)
+     */
+    funcdecl._scope = sc.copy();
+    funcdecl._scope.setNoFree();
+
+    __gshared bool printedMain = false; // semantic might run more than once
+    if (global.params.v.verbose && !printedMain)
+    {
+        const(char)* type = funcdecl.isMain() ? "main" : funcdecl.isWinMain() ? "winmain" : funcdecl.isDllMain() ? "dllmain" : cast(const(char)*)null;
+        Module mod = sc._module;
+
+        if (type && mod)
         {
-            //printf("\tnot virtual\n");
-            goto Ldone;
+            printedMain = true;
+            auto name = mod.srcfile.toChars();
+            auto path = FileName.searchPath(global.importPaths, name, true);
+            message("entry     %-10s\t%s", type, path ? path : name);
         }
-        // Suppress further errors if the return type is an error
-        if (funcdecl.type.nextOf() == Type.terror)
-            goto Ldone;
+    }
 
-        bool may_override = false;
-        for (size_t i = 0; i < cd.baseclasses.length; i++)
+    if (funcdecl.fbody && sc._module.isRoot() &&
+        (funcdecl.isMain() || funcdecl.isWinMain() || funcdecl.isDllMain() || funcdecl.isCMain()))
+        global.hasMainFunction = true;
+
+    if (funcdecl.fbody && funcdecl.isMain() && sc._module.isRoot())
+    {
+        // check if `_d_cmain` is defined
+        bool cmainTemplateExists()
+        {
+            Dsymbol pscopesym;
+            auto rootSymbol = sc.search(funcdecl.loc, Id.empty, pscopesym);
+            if (auto moduleSymbol = rootSymbol.search(funcdecl.loc, Id.object))
+                if (moduleSymbol.search(funcdecl.loc, Id.CMain))
+                    return true;
+
+            return false;
+        }
+
+        // Only mixin `_d_cmain` if it is defined
+        if (cmainTemplateExists())
         {
-            BaseClass* b = (*cd.baseclasses)[i];
-            ClassDeclaration cbd = b.type.toBasetype().isClassHandle();
-            if (!cbd)
+            // add `mixin _d_cmain!();` to the declaring module
+            auto tqual = new TypeIdentifier(funcdecl.loc, Id.CMain);
+            auto tm = new TemplateMixin(funcdecl.loc, null, tqual, null);
+            sc._module.members.push(tm);
+        }
+    }
+
+    assert(funcdecl.type.ty != Terror || funcdecl.errors);
+
+    // semantic for parameters' UDAs
+    foreach (i, param; f.parameterList)
+    {
+        if (param && param.userAttribDecl)
+            param.userAttribDecl.dsymbolSemantic(sc);
+    }
+}
+
+private int classFuncSemantic(ClassDeclaration cd, FuncDeclaration funcdecl,
```

Comment:
Needs documentation, what does it return and why?

Reply:
Done